### DISCUSSION TITLE
Added: LDM Notes for January 3rd, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5606] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-01-03.md

## Agenda

1. Slicing assumptions in list patterns, revisited
2. Parameterless struct constructors, revisited

**Comment by DavidArno** [comment]
> We will go with option 3, never synthesizing a parameterless constructor. If a struct has field initializers with no constructors, this is an error

So to clarify, will your code example fail to compile with this change?
```c#
Console.WriteLine(new S().field1); // This won't run as the code won't compile

struct S
{
    public int field1 = 1; // This is a compilation error as there's no default constructor
    public int field2;
    public S(int field2) => this.field2 = field2;
}
```

**Reply by cston** [reply]
> So to clarify, will your code example fail to compile with this change?

That example will compile because `S` includes an explicit constructor. But `new S().field1` will be `0` since there is no parameterless constructor. (Unless there is an explicit parameterless constructor, `new()` is equivalent to `default`.)

An error will be reported if there are field initializers and _no explicitly declared constructors_.
```csharp
struct S0 // error: 'struct' with field initializers must include an explicitly declared constructor
{
    public int field1 = 1;
}
```

**Reply by DavidArno** [reply]
But that's what it currently does, so I don't understand what the proposed breaking change is in that case. Surely the aim is to fix the issues with field initialisers and them not being invoked. I don't see how the proposed change addresses this. It seems to just ignore it and not even warn, which surely is the worst possible outcome?

**Reply by HaloFour** [reply]
Currently the compiler synthesizes the parameterless constructor if there are no other constructors defined, the breaking change will be to remove that and to have a compiler error that you need to define a constructor.

**Reply by DavidArno** [reply]
~So the proposal is to not only not fix the issue that constructors that take parameters don't call field initialisers nor warn of this fact, but to now require we add a pure-noise, empty, parameterless constructor to trigger the calling of those initialisers? What possible use are field initialisers in this scenario?~

~This decision is more than a little disappointing.~

Retracted as I was talking nonsense.

**Reply by HaloFour** [reply]
> So the proposal is to not only not fix the issue that constructors that take parameters don't call field initialisers

What do you mean?  Any declared constructor on the struct does run the field initializers first, that is the case today:

```cs
using System;

// prints 1
Console.WriteLine(new S(2).field1);

struct S
{
    public int field1 = 1;
    public int field2;
    public S(int field2) => this.field2 = field2;
}
```

**Reply by DavidArno** [reply]
Hmm, I've now realised how easy it is to not see any difference between `new S(2).field1` and `new S().field1` 😊

**Comment by huoyaoyuan** [comment]
I would take preference to consistency than compatibility. A silent behavior change is much worse than complaining for warning or error.

**Reply by CyrusNajmabadi** [reply]
That was our view as well.  With the updated, you should not get a silent behavior change.    It does mean you need to be more verbose, but we discussed this heavily and overwhelming voted to prefer that approach.  This wasn't the most desired outcome for some.  But it was felt to address the actual problems in this space, and was sensible and simple and effective in the timeframe we have *esp* as it's a breaking change.
